### PR TITLE
feat(Teamcity): Add Script to fetch aggregate test stats from TeamCity

### PR DESCRIPTION
**Pull request by rahulgurnani**
Presently, the script fetches the tests and the number of times they have failed.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)

### CODE REVIEW

File: teamcity/test_stats.go [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+	buildDataList := fetchAllBuildsSince(buildType, sinceString)
+
+	// Get the tests that ran on the last build
+	if len(buildDataList) == 0 {
+		log.Fatalln("No builds found")
+	}
+	ch := make(chan map[string]TestData)
+	go fetchTestsForBuild(buildDataList[0].ID, ch)
+	testsMap := <-ch
```
Comment (rahulgurnani) [inline teamcity/test_stats.go:183]: firstTestMap

File: teamcity/test_stats.go [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+		testStats.TotalRuns++
+		testStatsMap[k] = testStats
+	}
+	// Find the tests that fail the most percentage wise and output top 10
+	for i := 1; i < len(buildDataList); i++ {
+		go fetchTestsForBuild(buildDataList[i].ID, ch)
+	}
+	for i := 1; i < len(buildDataList); i++ {
+		currentTestsMap := <-ch
+		for k := range testsMap {
```
Comment (rahulgurnani) [inline teamcity/test_stats.go:202]: testName

File: teamcity/test_stats.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+			test, found := currentTestsMap[k]
+			if !found {
+				continue
+			}
+			var temp = testStatsMap[k]
```
Comment (rahulgurnani) [inline teamcity/test_stats.go:207]: naming

**Review by parasssh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)* status: 0 of 2 files reviewed, 7 unresolved discussions (waiting on @danielmai, @manishrjain, @parasssh, @rahulgurnani, and @vvbalaji-dgraph)

___
*[teamcity/test_stats.go, line 108 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIu6yJ29SDe4ZH5ZkBf:-MIu6yJ33cKSltqGYruE:b-gf5vg) ([raw file](https://github.com/dgraph-io/dgraph/blob/8737662d65a4b0939e5e8f53539ca13b42db61f1/teamcity/test_stats.go#L108)):*
> ```Go
> // Fetch the status of all the tests that ran for the given buildId
> func fetchTestsForBuild(buildID int, ch chan<- map[string]TestData) {
> 	url := fmt.Sprintf("https://teamcity.dgraph.io/app/rest/testOccurrences?locator=build:id:%d", buildID)
> ```

lets not hard code the URL. 
___
*[teamcity/test_stats.go, line 138 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIuBeNy9lISg6weviNU:-MIuBeNy9lISg6weviNV:b-4ff0ke) ([raw file](https://github.com/dgraph-io/dgraph/blob/8737662d65a4b0939e5e8f53539ca13b42db61f1/teamcity/test_stats.go#L138)):*
> ```Go
> }
> 
> func fetchAllBuildsSince(buildType string, date string) []BuildData {
> ```

this func returns slice. Prev function adds it to channel. Can we make it consistent?
___
*[teamcity/test_stats.go, line 140 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIuENT05Nqi7hpAPtc8:-MIuENT05Nqi7hpAPtc9:bup0mve) ([raw file](https://github.com/dgraph-io/dgraph/blob/8737662d65a4b0939e5e8f53539ca13b42db61f1/teamcity/test_stats.go#L140)):*
> ```Go
> func fetchAllBuildsSince(buildType string, date string) []BuildData {
> 	url := fmt.Sprintf("https://teamcity.dgraph.io/app/rest/builds/?locator=branch:refs/heads/master,buildType:%s,sinceDate:%s", buildType, date)
> 	url = strings.ReplaceAll(url, "+", "%2B")
> ```

maybe abstract all this in a URL builder type or something?
___
*[teamcity/test_stats.go, line 221 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIuEa0fAXBrX3pJ_ftf:-MIuEa0fAXBrX3pJ_ftg:b-3co2hq) ([raw file](https://github.com/dgraph-io/dgraph/blob/8737662d65a4b0939e5e8f53539ca13b42db61f1/teamcity/test_stats.go#L221)):*
> ```Go
> 		}
> 	}
> 	// Sort the test in ascending order of flakiness = failures / total runs
> ```

That is not the definition of flakiness. Flakiness are tests that are unreliable and so flaky tests are ones where failures/total_runs is > 20% and <80%.

However, outputing them in this order is fine.

**Review by rahulgurnani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)* status: 0 of 2 files reviewed, 7 unresolved discussions (waiting on @danielmai, @manishrjain, @parasssh, @rahulgurnani, and @vvbalaji-dgraph)

___
*[teamcity/test_stats.go, line 221 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIuEa0fAXBrX3pJ_ftf:-MJ1IpiREknfdZZ2f4sK:b-32c2o2) ([raw file](https://github.com/dgraph-io/dgraph/blob/8737662d65a4b0939e5e8f53539ca13b42db61f1/teamcity/test_stats.go#L221)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

That is not the definition of flakiness. Flakiness are tests that are unreliable and so flaky tests are ones where failures/total_runs is > 20% and <80%.

However, outputing them in this order is fine. 
</blockquote></details>

No. "flaky" test result as a test that exhibits both a passing and a failing result with the same code. https://testing.googleblog.com/2016/05/flaky-tests-at-google-and-how-we.html

**Review by rahulgurnani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @danielmai, @manishrjain, @parasssh, @rahulgurnani, and @vvbalaji-dgraph)

___
*[teamcity/test_stats.go, line 207 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIsfccD804yBZkSkacR-r1-207:-MJ1JBo_4XAcZxVZfVY_:bb6un1j) ([raw file](https://github.com/dgraph-io/dgraph/blob/fb5783bbd846015a369f491e3dedf87f68b47162/teamcity/test_stats.go#L207)):*
<details><summary><i>Previously, rahulgurnani (Rahul Gurnani) wrote…</i></summary><blockquote>

naming
</blockquote></details>

Done

**Review by rahulgurnani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @danielmai, @manishrjain, @parasssh, @rahulgurnani, and @vvbalaji-dgraph)

___
*[teamcity/test_stats.go, line 202 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIsfccD804yBZkSkacR-r1-202:-MJ1JFn-2ijK1-k-THfF:bb6un1j) ([raw file](https://github.com/dgraph-io/dgraph/blob/fb5783bbd846015a369f491e3dedf87f68b47162/teamcity/test_stats.go#L202)):*
<details><summary><i>Previously, rahulgurnani (Rahul Gurnani) wrote…</i></summary><blockquote>

testName
</blockquote></details>

Done

**Review by rahulgurnani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @danielmai, @manishrjain, @parasssh, @rahulgurnani, and @vvbalaji-dgraph)

___
*[teamcity/test_stats.go, line 183 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIsfccD804yBZkSkacR-r1-183:-MJ1JHOQ-cq1jyHbV5Xn:bb6un1j) ([raw file](https://github.com/dgraph-io/dgraph/blob/fb5783bbd846015a369f491e3dedf87f68b47162/teamcity/test_stats.go#L183)):*
<details><summary><i>Previously, rahulgurnani (Rahul Gurnani) wrote…</i></summary><blockquote>

firstTestMap
</blockquote></details>

Done

**Review by rahulgurnani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @danielmai, @manishrjain, @parasssh, @rahulgurnani, and @vvbalaji-dgraph)

___
*[teamcity/test_stats.go, line 138 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIuBeNy9lISg6weviNU:-MJ1b89G35NJHbJq7zcp:bp1lvy3) ([raw file](https://github.com/dgraph-io/dgraph/blob/8737662d65a4b0939e5e8f53539ca13b42db61f1/teamcity/test_stats.go#L138)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

this func returns slice. Prev function adds it to channel. Can we make it consistent?
</blockquote></details>

This function call is a blocking call. It's not required to create a channel here. The previous function is called concurrently so it required a channel to send the value back to.

**Review by rahulgurnani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @danielmai, @manishrjain, @parasssh, @rahulgurnani, and @vvbalaji-dgraph)

___
*[teamcity/test_stats.go, line 140 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIuENT05Nqi7hpAPtc8:-MJ1cME902j1pLCOODVl:ba0zzlc) ([raw file](https://github.com/dgraph-io/dgraph/blob/8737662d65a4b0939e5e8f53539ca13b42db61f1/teamcity/test_stats.go#L140)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

maybe abstract all this in a URL builder type or something?
</blockquote></details>

Not sure how to do that here

**Review by rahulgurnani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @danielmai, @manishrjain, @parasssh, and @vvbalaji-dgraph)

___
*[teamcity/test_stats.go, line 108 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIu6yJ29SDe4ZH5ZkBf:-MJ1cPZn1WJ_NZkRnIca:b5buyjg) ([raw file](https://github.com/dgraph-io/dgraph/blob/8737662d65a4b0939e5e8f53539ca13b42db61f1/teamcity/test_stats.go#L108)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

lets not hard code the URL. 
</blockquote></details>

Do you want me to move base URL to a constant and have a suffix that I construct here?

**Comment by parasssh** [comment]
> > _[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)_ status: 0 of 2 files reviewed, 7 unresolved discussions (waiting on @danielmai, @manishrjain, @parasssh, @rahulgurnani, and @vvbalaji-dgraph)
> 
> _[teamcity/test_stats.go, line 221 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIuEa0fAXBrX3pJ_ftf:-MJ1IpiREknfdZZ2f4sK:b-32c2o2) ([raw file](https://github.com/dgraph-io/dgraph/blob/8737662d65a4b0939e5e8f53539ca13b42db61f1/teamcity/test_stats.go#L221)):_
> Previously, parasssh wrote…
> 
> > That is not the definition of flakiness. Flakiness are tests that are unreliable and so flaky tests are ones where failures/total_runs is > 20% and <80%.
> > However, outputing them in this order is fine.
> 
> No. "flaky" test result as a test that exhibits both a passing and a failing result with the same code. https://testing.googleblog.com/2016/05/flaky-tests-at-google-and-how-we.html

Oh ok. So, if the code changes, then we dont know if the pass -> fail or fail -> pass transition for the same test is legit or flaky? Conversely, if the same code is run again and the result is pass and/or fail then it is flaky. 
How do we capture flaky ones then? Seems like a hard one.  (I didnt read the full blog)

**Review by parasssh — APPROVED** [review]

**Review by rahulgurnani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6650)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @danielmai, @manishrjain, @parasssh, and @vvbalaji-dgraph)

___
*[teamcity/test_stats.go, line 221 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6650#-MIuEa0fAXBrX3pJ_ftf:-MJaM17m9rxUNtOfkXOb:bklor0f) ([raw file](https://github.com/dgraph-io/dgraph/blob/8737662d65a4b0939e5e8f53539ca13b42db61f1/teamcity/test_stats.go#L221)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

Oh ok. So, if the code changes, then we dont know if the pass -> fail or fail -> pass transition for the same test is legit or flaky? Conversely, if the same code is run again and the result is pass and/or fail then it is flaky.  How do we capture flaky ones then? Seems like a hard one.  (I didnt read the full blog) 
</blockquote></details>

We can know if a test is flaky only by running it n (maybe = 100) times and see if it passes n times. That would tell us with a high probability that the test is non-flaky or it's indeed flaky.
But for simplicity, based on test stats we get from this tool, we should mark some of the tests as flaky. What we should do is retry the flaky ones for 3 times in the CI and then fail the CI. That way we would significantly reduce the probability of CI failing because of a flaky test.

**Pull request merged**